File: src/main/java/io/github/binout/soccer/interfaces/rest/SeasonMatchesResource.java
package io.github.binout.soccer.interfaces.rest;

import io.github.binout.soccer.application.season.*;
import io.github.binout.soccer.domain.date.FriendlyMatchDate;
import io.github.binout.soccer.domain.date.LeagueMatchDate;
import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.domain.season.match.Match;
import io.github.binout.soccer.interfaces.rest.model.RestDate;
import io.github.binout.soccer.interfaces.rest.model.RestMatch;
import io.vavr.Tuple2;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("rest/seasons/{name}/matches")
public class SeasonMatchesResource {

    @Autowired
    GetFriendlyMatches getFriendlyMatches;

    @Autowired
    GetNextFriendlyMatches getNextFriendlyMatches;

    @Autowired
    AddFriendlyMatch addFriendlyMatch;

    @Autowired
    GetLeagueMatches getLeagueMatches;

    @Autowired
    GetNextLeagueMatches getNextLeagueMatches;

    @Autowired
    AddLeagueMatch addLeagueMatch;

    @Autowired
    GetToPlanFriendlyMatches getToPlanFriendlyMatches;

    @Autowired
    GetToPlanLeagueMatches getToPlanLeagueMatches;

    @Autowired
    SubstitutePlayerInFriendlyMatches substitutePlayerInFriendlyMatches;

    @Autowired
    SubstitutePlayerInLeagueMatches substitutePlayerInLeagueMatches;

    @GetMapping("friendly")
    public ResponseEntity getFriendlyMatch(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getFriendlyMatches.execute(seasonName).map(this::toRestMatch).collect(Collectors.toList()));
    }

    @PutMapping("friendly/{dateParam}")
    public ResponseEntity putFriendlyMatch(@PathVariable("name") String name, @PathVariable("dateParam") String dateParam) {
        String seasonName = new SeasonName(name).name();
        RestDate date = new RestDate(dateParam);
        addFriendlyMatch.execute(seasonName, date.year(), date.month(), date.day());
        return ResponseEntity.ok().build();
    }

    @GetMapping("friendly/next")
    public ResponseEntity getNextFriendly(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getNextFriendlyMatches.execute(seasonName).map(this::toRestMatch).collect(Collectors.toList()));
    }

    @GetMapping("friendly/to-plan")
    public ResponseEntity friendlyToPlan(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getToPlanFriendlyMatches.execute(seasonName)
                        .map(FriendlyMatchDate::date)
                        .map(RestMatch::new)
                        .collect(Collectors.toList()));
    }

    @DeleteMapping("friendly/{dateParam}/players/{playerName}")
    public ResponseEntity susbstitutePlayerFriendly(@PathVariable("name") String name, @PathVariable("dateParam") String dateParam, @PathVariable("playerName") String playerName) {
        String seasonName = new SeasonName(name).name();
        RestDate date = new RestDate(dateParam);
        substitutePlayerInFriendlyMatches.execute(seasonName, date.asLocalDate(), playerName);
        return ResponseEntity.ok().build();
    }


    @GetMapping("league")
    public ResponseEntity getLeagueMatch(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getLeagueMatches.execute(seasonName).map(this::toRestMatch).collect(Collectors.toList()));
    }

    @PutMapping("league/{dateParam}")
    public ResponseEntity putLeagueMatch(@PathVariable("name") String name, @PathVariable("dateParam") String dateParam) {
        String seasonName = new SeasonName(name).name();
        RestDate date = new RestDate(dateParam);
        addLeagueMatch.execute(seasonName, date.year(), date.month(), date.day());
        return ResponseEntity.ok().build();
    }

    @GetMapping("league/next")
    public ResponseEntity getNextLeague(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getNextLeagueMatches.execute(seasonName).map(this::toRestMatch).collect(Collectors.toList()));
    }

    @GetMapping("league/to-plan")
    public ResponseEntity leagueToPlan(@PathVariable("name") String name) {
        String seasonName = new SeasonName(name).name();
        return ResponseEntity.ok(getToPlanLeagueMatches.execute(seasonName)
                .map(LeagueMatchDate::date)
                .map(RestMatch::new)
                .collect(Collectors.toList()));
    }

    @DeleteMapping("league/{dateParam}/players/{playerName}")
    public ResponseEntity susbstitutePlayerLeague(@PathVariable("name") String name, @PathVariable("dateParam") String dateParam, @PathVariable("playerName") String playerName) {
        String seasonName = new SeasonName(name).name();
        RestDate date = new RestDate(dateParam);
        substitutePlayerInLeagueMatches.execute(seasonName, date.asLocalDate(), playerName);
        return ResponseEntity.ok().build();
    }


    private RestMatch toRestMatch(Tuple2<? extends Match,List<Player>> m) {
        RestMatch restMatch = new RestMatch(m._1.date());
        m._1.players().forEach(restMatch::addPlayer);
        m._2.stream().map(Player::name).forEach(restMatch::addSub);
        return restMatch;
    }
}

File: src/main/kotlin/io/github/binout/soccer/interfaces/rest/SeasonMatchesResource.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.interfaces.rest

import io.github.binout.soccer.application.*
import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.player.values
import io.github.binout.soccer.domain.season.Match
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("rest/seasons/{name}/matches")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
class SeasonMatchesResource {

        @Inject lateinit var getFriendlyMatches: GetFriendlyMatches
        @Inject lateinit var getNextFriendlyMatches: GetNextFriendlyMatches
        @Inject lateinit var addFriendlyMatch: AddFriendlyMatch
        @Inject lateinit var getLeagueMatches: GetLeagueMatches
        @Inject lateinit var getNextLeagueMatches: GetNextLeagueMatches
        @Inject lateinit var addLeagueMatch: AddLeagueMatch
        @Inject lateinit var getToPlanFriendlyMatches: GetToPlanFriendlyMatches
        @Inject lateinit var getToPlanLeagueMatches: GetToPlanLeagueMatches
        @Inject lateinit var substitutePlayerInFriendlyMatches: SubstitutePlayerInFriendlyMatches
        @Inject lateinit var substitutePlayerInLeagueMatches: SubstitutePlayerInLeagueMatches

    @GET
    @Path("friendly")
    fun getFriendlyMatch(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getFriendlyMatches.execute(seasonName).map { this.toRestMatch(it) }).build()
    }

    @PUT
    @Path("friendly/{dateParam}")
    fun putFriendlyMatch(@PathParam("name") name: String, @PathParam("dateParam") dateParam: String): Response {
        val seasonName = SeasonName(name).name
        val date = dateParam.toRestDate()
        addFriendlyMatch.execute(seasonName, date.year, date.month, date.day)
        return Response.ok().build()
    }

    @GET
    @Path("friendly/next")
    fun getNextFriendly(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getNextFriendlyMatches.execute(seasonName).map{ this.toRestMatch(it) }).build()
    }

    @GET
    @Path("friendly/to-plan")
    fun friendlyToPlan(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getToPlanFriendlyMatches.execute(seasonName)
                .map { it.date.toRestMatch() })
                .build()
    }

    @DELETE
    @Path("friendly/{dateParam}/players/{playerName}")
    fun susbstitutePlayerFriendly(@PathParam("name") name: String, @PathParam("dateParam") dateParam: String, @PathParam("playerName") playerName: String): Response {
        val seasonName = SeasonName(name).name
        val date = dateParam.toRestDate()
        substitutePlayerInFriendlyMatches.execute(seasonName, date.asLocalDate(), playerName)
        return Response.ok().build()
    }


    @GET
    @Path("league")
    fun getLeagueMatch(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getLeagueMatches.execute(seasonName).map { this.toRestMatch(it) }).build()
    }

    @PUT
    @Path("league/{dateParam}")
    fun putLeagueMatch(@PathParam("name") name: String, @PathParam("dateParam") dateParam: String): Response {
        val seasonName = SeasonName(name).name
        val date = dateParam.toRestDate()
        addLeagueMatch.execute(seasonName, date.year, date.month, date.day)
        return Response.ok().build()
    }

    @GET
    @Path("league/next")
    fun getNextLeague(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getNextLeagueMatches.execute(seasonName).map { this.toRestMatch(it) }).build()
    }

    @GET
    @Path("league/to-plan")
    fun leagueToPlan(@PathParam("name") name: String): Response {
        val seasonName = SeasonName(name).name
        return Response.ok(getToPlanLeagueMatches.execute(seasonName).map { it.date.toRestMatch() }).build()
    }

    @DELETE
    @Path("league/{dateParam}/players/{playerName}")
    fun susbstitutePlayerLeague(@PathParam("name") name: String, @PathParam("dateParam") dateParam: String, @PathParam("playerName") playerName: String): Response {
        val seasonName = SeasonName(name).name
        val date = dateParam.toRestDate()
        substitutePlayerInLeagueMatches.execute(seasonName, date.asLocalDate(), playerName)
        return Response.ok().build()
    }

    private fun toRestMatch(m: Pair<Match<*>, List<Player>>): RestMatch = m.let { (match, players) ->
        val restMatch = match.date.toRestMatch()
        restMatch.hasMinimumPlayer = match.hasMinimumPlayer()
        match.players().values().forEach { restMatch.players.add(it) }
        players.map { it.name }.values().forEach { restMatch.subs.add(it) }
        return restMatch
    }
}
